README.txt
ECONOMY INC - 

RoadMap

[✔] FIX BILLS STORAGE
[✔] LIGHTING THE TESR MODELS (FIXED partially, no lightmap, so everything is bright, need to add a bit of light)
[✔] GEAR AND GEAR MECHANISM ITEMS
[✔] FIX RECIPES
[✔] PACKETS OF DIFFERENT BILLS
[✔] DIGITAL VAULT CRACKER (Partially)
[✔] BLOCK SELLER DISPLAYING ITEM 3D (Fixed but currently is glitching out for some reason)
[✔] GOLD NUGGET WEIGHT + REMOVE RECIPE (Fixed partially, still thinking about a fair method for gold exchanging into money)
[✔] GOLD CHANGER WORKING Fixed
[✔] COMMAND PLOTS Fixed
[✔] PLOTS SAVING (WORLD SAVED DATA) Fixed
[] BLOCK SAFE VAULT FOR VILLAGES - Suspended
[] VILLAGES IMPLEMENTATION OF BANK + TABLE LIST - Suspended
[] CREATE INFORMER VILLAGER - Suspended
[] MYSQL


changelog.txt
Build: 1.16.5-36.0.15 - Sun Feb 07 03:41:41 GMT 2021
	cyborgmas18:
		Fix IAngerable entities crashing when reading nbt on client (#7618)

=========
Build: 1.16.5-36.0.14 - Tue Feb 02 20:28:08 GMT 2021
	thesilkminer:
		[1.16.x] Add loot table ID in Global Loot Modifiers (#7428)

=========
Build: 1.16.5-36.0.13 - Sat Jan 30 02:07:17 GMT 2021
	richard:
		Add a block tag to allow mods to blacklist blocks endermen should not be able to place blocks on top of, similar to how vanilla hardcodes against them being able to place things on bedrock (#7548)

=========
Build: 1.16.5-36.0.12 - Sat Jan 30 02:01:35 GMT 2021
	7419378+hellfirepvp:
		Fix translucent model parts of MultiLayerModels rendering as missing models due to differing rendertypes when rendering block model (#7579)
		
		Signed-off-by: HellFirePvP <7419378+HellFirePvP@users.noreply.github.com>

=========
Build: 1.16.5-36.0.11 - Sat Jan 30 01:35:10 GMT 2021
	richard:
		Fix existing file tag provider support looking in the wrong folder for custom tag types (#7576)

=========
Build: 1.16.5-36.0.10 - Thu Jan 28 21:07:30 GMT 2021
	cyborgmas18:
		Propery re-encode registries (#7629)

=========
Build: 1.16.5-36.0.9 - Thu Jan 28 01:23:13 GMT 2021
	cyborgmas18:
		Properly encode registries. (#7626)

=========
Build: 1.16.5-36.0.8 - Thu Jan 28 00:00:43 GMT 2021
	cyborgmas18:
		Inject vanilla dimensions inside worlds missing them (#7599)

=========
Build: 1.16.5-36.0.7 - Wed Jan 27 00:32:53 GMT 2021
	lexmanos:
		Update Coremods, ASM, and include standalone Nashorn to load on J15.

	lexmanos:
		Introduce a new field_to_instanceof transformer to reduce some simple patches.

=========
Build: 1.16.5-36.0.4 - Tue Jan 26 02:53:10 GMT 2021
	lexmanos:
		Fix broken build.

	dev:
		Add modded biomes to overworld biome list (#7360)

=========
Build: 1.16.5-36.0.2 - Tue Jan 26 02:25:10 GMT 2021
	lexmanos:
		Fix dynamic registries not getting ids assigned correctly from old worlds in specific unordered value insertions. Closes #7586

=========
Build: 1.16.5-36.0.1 - Sun Jan 17 21:05:36 GMT 2021
	knightminer4:
		Optionally add a milk fluid to the vanilla bucket (#7294)
		
		Fixes the previous broken code that was supposed to make mod registered milk returned by the vanilla bucket by allowing mods to enable a milk fluid.

=========
Build: 1.16.5-36.0.0 - Fri Jan 15 19:42:24 GMT 2021
	lexmanos:
		1.16.5 Initial Update.

=========
Build: 1.16.4-35.1.37 - Thu Jan 14 23:58:35 GMT 2021
	lexmanos:
		Update Coremods, ASM, and include standalone Nashorn to load on J15.

=========
Build: 1.16.4-35.1.36 - Thu Jan 07 20:24:49 GMT 2021
	ian.rehwinkel:
		Added IWeatherParticleRenderHandler to allow a mod to take control over weather particles and sounds. (#7522)

=========
Build: 1.16.4-35.1.35 - Thu Jan 07 20:16:03 GMT 2021
	covers1624:
		Allow Chunk loading Tickets to opt-into forcing chunk ticks. (#7525)

=========
Build: 1.16.4-35.1.34 - Thu Jan 07 20:10:18 GMT 2021
	malte0811:
		Fix crash when mods add goals to skeleton horses (MC-206338) (#7509)

=========
Build: 1.16.4-35.1.33 - Thu Jan 07 20:03:59 GMT 2021
	diesieben07:
		Set WM_CLASS and WM_INSTANCE_NAME for early progress window (#7534)

=========
Build: 1.16.4-35.1.32 - Wed Jan 06 03:17:35 GMT 2021
	35673674+alcatrazescapee:
		Make UnboundedMapCodec more lenient in decoding dimensions. Fixes MC-197860 (#7527)

=========
Build: 1.16.4-35.1.31 - Tue Jan 05 19:59:52 GMT 2021
	lexmanos:
		Prevent NPE and deprecated MissingMappings.getMappings function. Closes #6252

	lexmanos:
		Fix FML's Config Sync packet not allowing same length filenames in encoder and decoder. Closes #7584

=========
Build: 1.16.4-35.1.29 - Sun Jan 03 11:02:59 GMT 2021
	lexmanos:
		Fix entities not correctly being able to be ridden under water.

=========
Build: 1.16.4-35.1.28 - Mon Dec 28 21:56:51 GMT 2020
	git:
		Fix crash when beehive is broken by fake player (#7566)

=========
Build: 1.16.4-35.1.27 - Mon Dec 28 21:46:04 GMT 2020
	35673674+alcatrazescapee:
		Add supplier variant to `AbstractBlock.Properties#lootFrom` (#7563)

=========
Build: 1.16.4-35.1.26 - Mon Dec 28 21:40:18 GMT 2020
	diesieben07:
		Fix missing serializers when deserializing global loot modifiers (#7561)

	diesieben07:
		Fix ModFileResourcePack.getAllResourceLocations ignoring resourceNamespace. (#7562)
		
		Fixes #7405

=========
Build: 1.16.4-35.1.24 - Mon Dec 28 21:34:39 GMT 2020
	diesieben07:
		Fix VanillaConnectionNetworkFilter sending out invalid entity properties packets without an entity ID (#7560)

=========
Build: 1.16.4-35.1.23 - Mon Dec 28 21:17:01 GMT 2020
	richard:
		Make shear interactions extend to subclasses (#7544)

=========
Build: 1.16.4-35.1.22 - Mon Dec 28 21:09:38 GMT 2020
	sciwhiz12:
		Add a SoundType subclass that uses suppliers (#7538)

=========
Build: 1.16.4-35.1.21 - Mon Dec 28 21:03:56 GMT 2020
	sciwhiz12:
		Fix compound ingredient in shapeless recipes (#7537) Fixes #7530

=========
Build: 1.16.4-35.1.20 - Mon Dec 28 20:19:04 GMT 2020
	matthrewp:
		Fix modded PointOfInterestType's not populating blockstate map when registered (#7503)

	35673674+alcatrazescapee:
		Fix potential NPE in ForgeHooks.onItemPlaceIntoWorld when passing in null player. (#7505)

=========
Build: 1.16.4-35.1.18 - Mon Dec 28 20:13:15 GMT 2020
	48399898+poopoodice:
		Add new LivingConversionEvent to control mobs turning into other mobs. (#7386)

=========
Build: 1.16.4-35.1.17 - Mon Dec 28 20:01:42 GMT 2020
	knightminer4:
		Add ItemAttributeModifierEvent (#7484)

=========
Build: 1.16.4-35.1.16 - Mon Dec 28 19:54:38 GMT 2020
	bward7864:
		Fix RightClickBlock ALLOW not being implemented (#7426)

=========
Build: 1.16.4-35.1.15 - Sat Dec 26 19:39:35 GMT 2020
	lexmanos:
		Make mod data packs able to be disabled. It'll break things, but that's on you. Closes #6776

	lexmanos:
		Quiet down errors related to modders not changing their update urls from the default.

=========
Build: 1.16.4-35.1.13 - Sat Dec 12 17:34:07 GMT 2020
	tterrag:

src/main/java/fr/fifoube/gui/GuiSeller.java
/*******************************************************************************
 *******************************************************************************/
package fr.fifoube.gui;

import java.awt.Color;

import org.lwjgl.opengl.GL11;

import com.mojang.blaze3d.matrix.MatrixStack;
import com.mojang.blaze3d.systems.RenderSystem;

import fr.fifoube.blocks.tileentity.TileEntityBlockSeller;
import fr.fifoube.gui.container.ContainerSeller;
import fr.fifoube.gui.widget.RefillIconButton;
import fr.fifoube.main.ModEconomyInc;
import fr.fifoube.packets.PacketSellerCreated;
import fr.fifoube.packets.PacketsRegistery;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.screen.ConfirmScreen;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.screen.inventory.ContainerScreen;
import net.minecraft.client.gui.widget.TextFieldWidget;
import net.minecraft.client.gui.widget.button.Button;
import net.minecraft.client.gui.widget.button.LockIconButton;
import net.minecraft.client.resources.I18n;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.entity.player.PlayerInventory;
import net.minecraft.item.Items;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.StringTextComponent;
import net.minecraft.util.text.TranslationTextComponent;

public class GuiSeller extends ContainerScreen<ContainerSeller> {

	private static final ResourceLocation background = new ResourceLocation(ModEconomyInc.MOD_ID , "textures/gui/container/gui_seller.png");
	private TileEntityBlockSeller tile;
	protected int xSize = 176;
	protected int ySize = 168;
	protected int guiLeft;
	protected int guiTop;
	
	private Button validate;
	private Button unlimitedStack;
	private RefillIconButton autoRefill;
	private TextFieldWidget costField;

	private double cost;
	private boolean admin = false;
	
	private boolean validCost = false;
	
	public GuiSeller(ContainerSeller container, PlayerInventory playerInventory, ITextComponent name) 
	{
		super(container, playerInventory, name);
		this.tile = getContainer().getTile();
	}
	
	
	@Override
	public void tick() 
	{
		super.tick();
		this.validCost = checkIfTextCanBeParsedToPositiveDouble(this.costField.getText());
	}
	
	@Override
	protected void init() {
		
		super.init();
		this.fieldInit();
		PlayerEntity player = minecraft.player;
		if(!tile.getCreated())
		{
            this.autoRefill = this.addButton(new RefillIconButton(width / 2 + 87, height / 2 - 75, (onPress) -> { actionPerformed(this.autoRefill); }, background)); 
            this.validate = this.addButton(new Button(width / 2 + 26, height / 2 + 83, 55, 20, new TranslationTextComponent("title.validate"), (onPress) -> { actionPerformed(this.validate);}));
			if(player.isCreative() == true)
			{
				this.unlimitedStack = this.addButton(new Button(width /2 + 2, height / 2 - 96, 80, 13, new TranslationTextComponent(""), (onPress) -> { actionPerformed(this.unlimitedStack); }));
			}
		}
	}
	
	protected void fieldInit()
	{
	      this.minecraft.keyboardListener.enableRepeatEvents(true);
	      int i = (this.width - this.xSize) / 2;
	      int j = (this.height - this.ySize) / 2;
	      this.costField = new TextFieldWidget(this.font, i + 121, j + 15, 38, 12, new TranslationTextComponent("title.cost"));
	      this.costField.setCanLoseFocus(false);
	      this.costField.setTextColor(-1);
	      this.costField.setDisabledTextColour(-1);
	      this.costField.setEnableBackgroundDrawing(false);
	      this.costField.setMaxStringLength(35);
	      this.children.add(this.costField);
	      this.setFocusedDefault(this.costField);
	}
	
	@Override
	public void resize(Minecraft minecraft, int width, int height) {
		
		String s = this.costField.getText();
		this.init(minecraft, width, height);
		this.costField.setText(s);
	}
	
	@Override
	public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
		      if (keyCode == 256) {
		         this.minecraft.player.closeScreen();
		      }

		      return !this.costField.keyPressed(keyCode, scanCode, modifiers) && !this.costField.canWrite() ? super.keyPressed(keyCode, scanCode, modifiers) : true;
	}
	   
	protected void actionPerformed(Button button)
	{
		PlayerEntity playerIn = minecraft.player;
		if(tile != null)
		{
			if(button == this.unlimitedStack)
			{
				if(this.admin == false)
				{
					this.admin = true;
					tile.setAdmin(true);
				}
				else if(this.admin == true)
				{
					this.admin = false;
					tile.setAdmin(false);
				}
			}
			else if(button == this.autoRefill)
			{
				if(this.autoRefill.isRefillEnabled())
				{
					this.autoRefill.setRefillMode(false);
				}
				else
				{
					this.autoRefill.setRefillMode(true);
				}
			}
			else if(button == this.validate)
			{
				if(this.validCost)
				{
					this.cost = Double.valueOf(this.costField.getText());
					tile.setCost(Double.valueOf(this.costField.getText()));
					if(!(tile.getCost() == 0)) // IF TILE HAS NOT A COST OF 0 THEN WE PASS TO THE OTHER
					{
						if(tile.getStackInSlot(0).getItem() != Items.AIR) // IF SLOT 0 IS NOT BLOCKS.AIR, WE PASS
						{
							if(!this.admin) //ADMIN HASN'T SET : UNLIMITED STACK
							{
								tile.setAdmin(false);
								tile.setCreated(true); // CLIENT SET CREATED AT TRUE
								final int x = tile.getPos().getX(); // GET X
								final int y = tile.getPos().getY(); // GET Y
								final int z = tile.getPos().getZ(); // GET Z
								int amount = tile.getStackInSlot(0).getCount(); // GET COUNT IN TILE THANKS TO STACK IN SLOT
								String name = tile.getStackInSlot(0).getDisplayName().getString(); // GET ITEM NAME IN TILE THANKS TO STACK IN SLOT
								tile.setItem(name); // CLIENT SET ITEM NAME
								tile.setAutoRefill(this.autoRefill.isRefillEnabled());
								tile.markDirty();
								PacketsRegistery.CHANNEL.sendToServer(new PacketSellerCreated(true, this.cost, name, amount, x, y, z, false, tile.getAutoRefill())); // SEND SERVER PACKET FOR CREATED, COST, NAME, AMOUNT, X,Y,Z ARE TILE COORDINATES
								playerIn.closeScreen(); // CLOSE SCREEN
							}
							else if(this.admin) //ADMIN HAS SET : UNLIMITED STACK
							{
								tile.setAdmin(true);
								tile.setCreated(true); // CLIENT SET CREATED AT TRUE
								final int x = tile.getPos().getX(); // GET X
								final int y = tile.getPos().getY(); // GET Y
								final int z = tile.getPos().getZ(); // GET Z
								int amount = tile.getStackInSlot(0).getCount(); // GET COUNT IN TILE THANKS TO STACK IN SLOT
								String name = tile.getStackInSlot(0).getDisplayName().getString(); // GET ITEM NAME IN TILE THANKS TO STACK IN SLOT
								tile.setItem(name); // CLIENT SET ITEM NAME
								tile.setAutoRefill(this.autoRefill.isRefillEnabled());
								tile.markDirty();
								PacketsRegistery.CHANNEL.sendToServer(new PacketSellerCreated(true, this.cost, name, amount, x, y, z, true, tile.getAutoRefill())); // SEND SERVER PACKET FOR CREATED, COST, NAME, AMOUNT, X,Y,Z ARE TILE COORDINATES
								playerIn.closeScreen(); // CLOSE SCREEN
							}
							
						}
						else // PROVIDE PLAYER TO SELL AIR
						{
							playerIn.sendMessage(new StringTextComponent(I18n.format("title.sellAir")), playerIn.getUniqueID());	
						}
					}
					else // IT MEANS THAT PLAYER HAS NOT SELECTED A COST
					{
						playerIn.sendMessage(new StringTextComponent(I18n.format("title.noCost")), playerIn.getUniqueID());
					}
				}
				else
				{
					playerIn.sendMessage(new StringTextComponent(I18n.format("title.noValidCost")), playerIn.getUniqueID());
				}
			}
		}
	}
	
	@Override
		public boolean isPauseScreen() {
			return false;
		}
	
	@Override
	protected void drawGuiContainerForegroundLayer(MatrixStack matrixStack, int mouseX, int mouseY) 
	{  
		if(!tile.getCreated())
		{
			this.font.drawStringWithShadow(matrixStack, this.admin ? I18n.format("title.unlimited") :  I18n.format("title.limited"), this.xSize - 82, this.ySize - 177,  Color.WHITE.getRGB());
		}
		this.font.drawString(matrixStack, I18n.format("title.block_seller"), 8.0F, 5, Color.DARK_GRAY.getRGB());
	    this.font.drawString(matrixStack, this.playerInventory.getDisplayName().getString(), 8.0F, (float)(this.ySize - 94), 4210752);
	}
	
	
	@Override
	public void render(MatrixStack matrixStack, int mouseX, int mouseY, float partialTicks)
	{
		this.renderBackground(matrixStack);
		super.render(matrixStack, mouseX, mouseY, partialTicks);
		RenderSystem.disableBlend();
		if(!tile.getCreated())
		{
			this.costField.render(matrixStack, mouseX, mouseY, partialTicks);
		}
        this.renderHoveredTooltip(matrixStack, mouseX, mouseY);
	}
	
	@Override
	protected void drawGuiContainerBackgroundLayer(MatrixStack matrixStack, float partialTicks, int mouseX, int mouseY) 
	{
		
	    this.minecraft.getTextureManager().bindTexture(background); 
	    int k = (this.width - this.xSize) / 2; 
	    int l = (this.height - this.ySize) / 2;
	    this.blit(matrixStack, k, l, 0, 0, this.xSize, this.ySize); 
	    this.blit(matrixStack, k + 117, l + 11, 0, this.ySize + 32, 110, 16);
		if(!tile.getCreated())
		{
			this.blit(matrixStack, k + 117, l + 11, 0, this.ySize + (this.validCost ? 0 : 16), 110, 16);
		}

	}
	
	private boolean checkIfTextCanBeParsedToPositiveDouble(String field)
	{
		if(field != null) {
			try
			{	
				double value = Double.parseDouble(field);
				if(value > 0)
				{
					return true;
				}
				return false;
			}
			catch(NumberFormatException e)
			{
			  return false;
			}
		}
		return false;
	}
}
